rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,765 @@
+token instead of a secret key, and save the MFA token as a web cookie to avoid
+redirecting the user in the browser.
+
+#### SSO MFA session
+
+When a user requests an SSO MFA challenge, an SSO MFA session data object will
+be created in the backend. Once the user completes the SSO MFA flow, the client
+and session data will gain a matching MFA token, which can be used by the client
+to verify itself as the owner of the session.
+
+Note: challenge extensions are included, so scopes and reuse are both available.
+See RFD 155 for details.
+
+```go
+// SSOMFASessionData SSO MFA Session data.
+type SSOMFASessionData struct {
+  // RequestID is the ID of the corresponding SSO Auth request, which is used to
+  // identity this session.
+  RequestID string `json:"request_id,omitempty"`
+  // Username is the Teleport username.
+  Username string `json:"username,omitempty"`
+  // MFAToken is a secret token set when the SSO flow is completed and returned to
+  // the user. The user can then use this token to prove they own the MFA session.
+  MFAToken string `json:"token,omitempty"`
+  // ConnectorID is id of the corresponding Auth connector.
+  ConnectorID string `json:"connector_id,omitempty"`
+  // ConnectorType is SSO type of the corresponding Auth connector (SAML, OIDC, or Github).
+  ConnectorType string `json:"connector_type,omitempty"`
+  // ChallengeExtensions are challenge extensions that apply to this SSO MFA session.
+  ChallengeExtensions *mfav1.ChallengeExtensions `json:"challenge_extensions,omitempty"`
+}
+```
+
+#### MFA device
+
+Teleport will treat any user that originated from an mfa-enabled auth connector
+as if they have an SSO MFA device registered. This device is not stored in the
+backend. Instead, it is retrieved by checking the user's `CreatedBy` field,
+fetching the corresponding auth connector, checking if it has MFA enabled, and
+then filling out an SSO MFA device from the auth connector's details.
+
+### Proto
+
+**CreateAuthenticateChallengeRequest**
+
+```diff
+message CreateAuthenticateChallengeRequest {
+  ...
++ // SSOClientRedirectURL should be supplied If the client supports SSO MFA checks.
++ // If unset, the server will only return non-SSO challenges, or an error if SSO
++ // challenges are required.
++ string SSOClientRedirectURL = 7 [(gogoproto.jsontag) = "sso_client_redirect_url,omitempty"];
+}
+```
+
+**SSOChallenge**
+
+```proto
+// SSOChallenge contains SSO auth request details to perform an SSO MFA check.
+message SSOChallenge {
+  // RequestId is the ID of an SSO auth request.
+  string request_id = 1;
+  // RedirectUrl is an IdP redirect URL to initate the SSO MFA flow.
+  string redirect_url = 2;
+}
+```
+
+```diff
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
++   SSOResponse SSO = 4;
+  }
+}
+```
+
+**SSOResponse**
+
+```proto
+// SSOResponse is a response to SSOChallenge.
+message SSOResponse {
```

Comment:
nit: SSOChallengeResponse

SSOResponse could mean an actual SSO authn response, let's be specific.